### DISCUSSION TITLE
Better `unsafe` modifier support on partial methods

### DISCUSSION THREAD

**Discussion by AArnott** [dotnet/csharplang#7298] [Language Ideas]
The following C# is invalid, unexpectedly:

```cs
unsafe partial class A {
    partial void Foo(void* p);
}

partial class A {
    unsafe partial void Foo(void* p)
    {
        throw new System.NotImplementedException();
    }
}
```

[Full repro with error here on sharplab](https://sharplab.io/#v2:C4LghgzsA0AmIGoA+BXAdhMAzApgAgAcwAnYASzABs8ABAJjwEE8BvAWACg9vCTyraAFjwAxAPZiAFDUEAqQgEoA3JwC+nTkVIVq9Jq04886TLl7aBM0ROlzFhnuy5GewABbExAdzxocPmgBGADoAOTFgAEkAWwJKHGicNGAcWABRAA8AYxwCcjE0SWUHbnUOMqA).

The issue is the original partial class has `unsafe` on the class declaration, so the partial method doesn't need to have it directly.
But then the user's partial class that implements the partial method is not allowed to specify `unsafe` on the partial method. This is particularly ironic because the C# language service doesn't realize this when it auto-completes. When the user types `partial Foo` and hits Tab, C# produces this:

```cs
partial unsafe void Foo(void* p)
{
    throw new System.NotImplementedException();
}
```

This then produces *two* errors: one because `unsafe` isn't allowed, and one because `partial` must immediately precede the method return type. 
Through some trial and error, I found that the only way for the user to get compilable code with this partial method is to move the `unsafe` to the class declaration to match the source generated code:

```cs
unsafe partial class A {
    partial void Foo(void* p)
    {
        throw new System.NotImplementedException();
    }
}
```

So I have a couple ideas/requests:
1. The language service should produce valid code for auto-complete. This is probably a bug I should file at dotnet/roslyn.
2. Allow the `unsafe` modifier to appear on either partial method when the counterpart partial method _or its class declaration_ uses the `unsafe` modifier. This may be a language spec change.

**Comment by CyrusNajmabadi** [comment]
> This is probably a bug I should file at dotnet/roslyn.

Yes. :-)

> This may be a language spec change.

Yes. And this would be a big departure from current `unsafe` (which is purely lexical). It's possible, but it has really not be something we've felt a need to do up till now (esp. as the trivial fix is to put the modifier on the other partial type). But it can certainly be looked into.

**Reply by AArnott** [reply]
> the trivial fix is to put the modifier on the other partial type

Yes, doing it is trivial. Discovering it is less so. I spent a couple minutes wondering just how broken C# was since `unsafe` is both required and forbidden before realizing I could promote the `unsafe` to the class level. 
I guess we can keep punting this until more customer feedback comes on this. I _doubt_ CsWin32's exposing of this partial method will be used by hardly anyone, so it's probably not a big deal.

I'll file the lang svc bug.

**Reply by AArnott** [reply]
https://github.com/dotnet/roslyn/issues/68805

**Reply by CyrusNajmabadi** [reply]
>  I spent a couple minutes wondering just how broken C# was since unsafe is both required and forbidden before realizing I could promote the unsafe to the class level.

Generally speaking, i would classify that as a 'diagnostic clarity' issue.  Specifically, the diagnostic should both be clear as to what the problem is, and *ideally* guide people toward the appropriate fix.

**Comment by 333fred** [comment]
> Allow the unsafe modifier to appear on either partial method when the counterpart partial method or its class declaration uses the unsafe modifier. This may be a language spec change.

If we were going to make a change here, the most I'd want to do is allow `unsafe` to be missing from the partial declaration side, and only if the signature itself didn't contain anything that normally requires `unsafe` to be there (ie, if you take a pointer type as a parameter, `unsafe` should be required on both declaration and definition). As Cyrus mentioned, `unsafe` is lexical, and it feels extremely wrong for it to be lexically missing when there's a pointer type right there.

**Reply by AArnott** [reply]
> it feels extremely wrong for it to be lexically missing when there's a pointer type right there.

I'm not arguing for dropping it. I'm arguing for _allowing_ it, where today it is forbidden though required.

**Reply by AArnott** [reply]
In particular, since the signature clearly requires it, C# is crazy to disallow it, IMO. And then there's valid uses of `unsafe` modifier when the signature does not require it but the body does (as an alternative to creating an unsafe block within the method body). In the latter case, this is an implementation detail of the partial method definition, so adding it where the declaration didn't need it seems like it should be allowable.

**Reply by 333fred** [reply]
Ah, I see now. I read too fast initially. This is an interesting case, but I'd definitely need to think more on how the rules should evolve.

**Reply by RikkiGibson** [reply]
At the very least it feels like the diagnostic experience could be improved here.

It feels like if we are going to consider the unsafe modifier to apply *purely lexically* then there's also not a deep need for it to match across partial declarations. If the *lexical* unsafe context permits each declaration, we're fine.

We've separately gotten requests to permit only the implementation part to have `unsafe` in order to avoid needing the top-level `unsafe { }` block in the method. @jkoritzinsky may have filed that?

**Reply by CyrusNajmabadi** [reply]
Also, *personally* i think it would be *fine* to allow `unsafe` lexically here.

**Reply by jkoritzinsky** [reply]
Yep we opened https://github.com/dotnet/csharplang/discussions/6650 after a conversation in LDM a while back.
